src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -498,5 +498,75 @@ private static bool ShouldOptimizeOutInitializer(BoundStatement initializer)
             BoundExpression rhs = assignment.Right;
             return rhs.IsDefaultValue();
         }
+
+        /// <summary>
+        /// Receivers of struct methods are alowed to be RValues or LValues.
+        /// Whether the mutations from the method are propagated back to the 
+        /// receiver instance is conditional on whether the receiver is an LValue.
+        /// 
+        /// An inconvenient situation may arise when an the receiver is an RValue expression,
+        /// which is trivially reduced during lowering to one of its operands and 
+        /// such operand happens to be an LValue. That operation alone would 
+        /// expose the operand to mutations while it would not be exposed otherwise.
+        /// I.E. the transformation becomes semantically observable.
+        /// 
+        /// To prevent such situations, we will wrap the operand into a node whose only 
+        /// purpose is to never be an LValue.
+        /// </summary>
+        private static BoundExpression EnsureNotLvalueReceiver(BoundExpression expr)
+        {
+            // Leave as-is where receiver mutations cannot happen.
+            if (!ReceiverCanBeAssigned(expr))
+            {
+                return expr;
+            }
+
+            // - reference type receivers are byval
+            // - special value types do not have mutating members
+            var type = expr.Type.OriginalDefinition;
```

Comment:
Right. I have moved the check to the common helper, but forgot to remove it here.

---
In reply to: [108241731](https://github.com/dotnet/roslyn/pull/18166#discussion_r108241731) [](ancestors = 108241731)